### PR TITLE
docs(eslint-plugin): correct issues in documentation

### PR DESCRIPTION
**Pull request by armano2**
- description in documentation should match rule metadata
- validate if header in docs matches ``"# {description} (`{name}`)"``
- fix link to docs for some rules

i was also thinking about replacing this custom validation with jest test case

<details><summary>Test file</summary>
<p>
This code does same checks as `validate-docs`

```ts
import { promisify } from 'util';
import fs from 'fs';
import path from 'path';

import marked from 'marked';
import rules from '../src/rules';

// Promise api of fs is available for node 10+, for node 8 we have to use util
const existsAsync = promisify(fs.exists);
const readFileAsync = promisify(fs.readFile);
const readdirAsync = promisify(fs.readdir);

const docsRoot = path.resolve(__dirname, '../docs/rules');
const rulesData = Object.entries(rules);

function createRuleLink(ruleName: string): string {
  return `[\`@typescript-eslint/${ruleName}\`](./docs/rules/${ruleName}.md)`;
}

function parseReadme(): marked.Tokens.Table {
  const readmeRaw = fs.readFileSync(
    path.resolve(__dirname, '../README.md'),
    'utf8',
  );
  const readme = marked.lexer(readmeRaw, {
    gfm: true,
    silent: false,
  });

  // find the table
  const rulesTable = readme.find(
    token => token.type === 'table',
  ) as marked.Tokens.Table;
  if (!rulesTable) {
    throw Error('Could not find the rules table in README.md');
  }

  return rulesTable;
}

describe('Validating rule docs', () => {
  it(`check if all files are present`, async () => {
    const files = await readdirAsync(docsRoot);
    const ruleFiles = Object.keys(rules)
      .map(rule => `${rule}.md`)
      .sort();

    expect(files.sort()).toEqual(ruleFiles);
  });

  it.each(rulesData)(
    'Description of %s.md must match',
    async (ruleName, rule) => {
      const filePath = path.join(docsRoot, `${ruleName}.md`);
      expect(await existsAsync(filePath)).toBeTruthy();
      // validate if description of rule is same as in docs
      const file = await readFileAsync(filePath, 'utf-8');
      expect((/^\s*#[^\r\n]*/.exec(file) ?? [])[0]).toEqual(
        `# ${rule.meta.docs.description} (\`${ruleName}\`)`,
      );
    },
  );
});

describe('Validating rule metadata', () => {
  describe.each(rulesData)('%s', (ruleName, rule) => {
    it(`name field in rule must be correct`, () => {
      // validate if rule name is same as url
      // there is no way to access this field but its used only in generation of docs url
      expect(rule.meta.docs.url.endsWith(`rules/${ruleName}.md`)).toBeTruthy();
    });

    it('Expects type checking to be set for rule', async () => {
      // quick-and-dirty check to see if it uses parserServices
      // not perfect but should be good enough
      const ruleFileContents = await readFileAsync(
        path.resolve(__dirname, `../src/rules/${ruleName}.ts`),
      );

      const usesTypeInformation =
        ruleFileContents.includes('getParserServices') || undefined;
      expect(usesTypeInformation).toEqual(rule.meta.docs.requiresTypeChecking);
    });
  });
});

describe('Validating README.md', () => {
  const rulesTable = parseReadme().cells;
  const notDeprecated = rulesData.filter(
    ([, rule]) => rule.meta.deprecated !== true,
  );

  it('Checking table structure...', () => {
    const ruleNames = notDeprecated
      .map(([ruleName]) => ruleName)
      .sort()
      .map(createRuleLink);

    expect(rulesTable.map(row => row[0])).toStrictEqual(ruleNames);
  });

  describe.each(notDeprecated)(`Checking rule %s`, (ruleName, rule) => {
    const ruleRow =
      rulesTable.find(row => row[0].includes(`/${ruleName}.md`)) ?? [];

    it(`Link column should be set correctly`, () => {
      expect(ruleRow[0]).toEqual(createRuleLink(ruleName));
    });

    it(`Description column should be set correctly`, () => {
      expect(ruleRow[1]).toEqual(rule.meta.docs.description);
    });

    it(`Recommended column should be set correctly`, () => {
      expect(ruleRow[2]).toEqual(
        rule.meta.docs.recommended ? ':heavy_check_mark:' : '',
      );
    });

    it(`Fixable column should be set correctly`, () => {
      expect(ruleRow[3]).toEqual(
        rule.meta.fixable !== undefined ? ':wrench:' : '',
      );
    });

    it(`Requiring type information column should be set correctly`, () => {
      expect(ruleRow[4]).toEqual(
        rule.meta.docs.requiresTypeChecking === true ? ':thought_balloon:' : '',
      );
    });
  });
});
```

</p>
</details>

---------------

Q: How header in documentation for rule should look like? (right now we are its random)
- ``# {description} (`{name}`)``
- ``# {description} ({name})``

### CODE REVIEW

File: packages/eslint-plugin/src/rules/no-unnecessary-condition.ts [hunk 1]
```diff
@@ -72,7 +72,7 @@ export type MessageId =
   | 'never'
   | 'neverOptionalChain';
 export default createRule<Options, MessageId>({
-  name: 'no-unnecessary-conditionals',
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/no-unnecessary-condition.ts]: `no-unnecessary-conditionals` was invalid

File: packages/eslint-plugin/src/rules/no-magic-numbers.ts [hunk 2]
```diff
@@ -17,7 +17,7 @@ export default util.createRule<Options, MessageIds>({
   meta: {
     type: 'suggestion',
     docs: {
-      description: 'Disallows magic numbers',
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/no-magic-numbers.ts]: https://eslint.org/docs/rules/no-magic-numbers

File: packages/eslint-plugin/src/rules/no-untyped-public-signature.ts [hunk 3]
```diff
@@ -9,11 +9,10 @@ type MessageIds = 'noReturnType' | 'untypedParameter';
 type Options = [{ ignoredMethods: string[] }];
 
 export default util.createRule<Options, MessageIds>({
-  name: 'no-unused-public-signature',
```
Comment (armano2) [inline packages/eslint-plugin/src/rules/no-untyped-public-signature.ts]: next error `unused` => `untyped`

File: packages/eslint-plugin/docs/rules/return-await.md [hunk 4]
```diff
@@ -1,4 +1,4 @@
-# Require/Disallow returning awaited values in specific contexts (return-await)
+# Enforces consistent returning of awaited values in specific contexts (`return-await`)
```
Comment (armano2) [inline packages/eslint-plugin/docs/rules/return-await.md]: this sentence sound weird

Comment (bradzacher) [reply packages/eslint-plugin/docs/rules/return-await.md]: i wouldn't worry about "in specific contexts", I think it describes well enough without that.

File: packages/eslint-plugin/tools/validate-docs/check-for-rule-docs.ts [hunk 5]
```diff
@@ -2,23 +2,48 @@ import { TSESLint } from '@typescript-eslint/experimental-utils';
...
-      `Couldn't find file docs/rules/${ruleName}.md`,
-    );
+    if (!ruleHasDoc) {
+      errors.push(`Couldn't find file docs/rules/${ruleName}.md`);
+    } else {
+      const file = fs.readFileSync(
+        path.join(docsRoot, `${ruleName}.md`),
+        'utf-8',
+      );
+      const match = /^\s*#[^\r\n]*/.exec(file) ?? [];
```
Comment (bradzacher) [inline packages/eslint-plugin/tools/validate-docs/check-for-rule-docs.ts]: could do this via the markdown parser if you want something a bit more robust.

File: packages/eslint-plugin/tools/validate-docs/check-for-rule-docs.ts [hunk 6]
```diff
@@ -2,23 +2,48 @@ import { TSESLint } from '@typescript-eslint/experimental-utils';
...
+      const expectedDescription = `# ${rule.docs.description} (\`${ruleName}\`)`;
+      if (!match[0] || match[0] !== expectedDescription) {
+        errors.push(
+          'Header of file does not match the rule metadata.',
+          `    Expected: ${chalk.underline(expectedDescription)}`,
+          `    Received: ${chalk.underline(match[0])}`,
+        );
+      }
```
Comment (bradzacher) [inline packages/eslint-plugin/tools/validate-docs/check-for-rule-docs.ts]: i would probably prefer splitting the `!match[0]` case out, and explicitly state "No header of file", so things are a bit clearer (not that this should ever happen)

**Comment by bradzacher** [comment]
Thanks for doing this!
a while back I started writing a validator for the rule docs, but it was too much work at the time.
This is a great interim step!

**Comment by armano2** [comment]
ok, i applied requested changes,

bdw. what do you think about my idea about moving this out to jest test?

**Comment by bradzacher** [comment]
Sorry, I missed the description.

> How header in documentation for rule should look like? 

You've done it right - ``# description (`rule-name`)`` is the best way.

> bdw. what do you think about my idea about moving this out to jest test?

I actually don't mind the idea of moving this to a jest test. It's weird, I've written scripts like this so many times but I never thought about using a testing framework to report.

Does it work well?

**Comment by armano2** [comment]
it works in same way as any other test, but yest result is same and messages are ~simmilar

running it with other tests is actually faster (+~200ms) as opposed to ~4sec, running it in standalone mode `jest tests/docs.test.ts --runTestsByPath --silent` take same time.

**Pull request merged**